src/data_conn.c

```diff
@@ -1706,10 +1722,46 @@ uzfs_zvol_rebuild_scanner(void *arg)
 						    zinfo->name);
 						goto exit;
 					}
+					mutex_enter(&zv->rebuild_mtx);
+					if (zinfo->is_snap_inprogress) {
+						mutex_exit(&zv->rebuild_mtx);
+						sleep(1);
+						LOG_INFO("waiting for snapshot"
+						    " to finish");
+						goto retry;
+					}
+					zinfo->is_afs_inprogress = 1;
+					mutex_exit(&zv->rebuild_mtx);
 					uzfs_zvol_send_zio_cmd(zinfo, &hdr,
 					    ZVOL_OPCODE_REBUILD_ALL_SNAP_DONE,
 					    fd, NULL, 0, 0, warg);
 					all_snap_done = B_TRUE;
+					/*
+					 * wait for the downgraded replica to
+					 * transition into the AFS mode. After
+					 * that we can reset the afs_inprogress
+					 * as downgraded replica can now handle
+					 * the snapshot command gracefully.
+					 */
+					rc = uzfs_zvol_read_header(fd, &hdr);
+					if (rc != 0) {
+						LOG_ERR("afs started read"
+						    " failed zvol %s err(%d)",
+						    zinfo->name, rc);
+						goto exit;
```

Comment:
not resetting is_afs_inprogress here, which is kind of bug, can be avoided by moving this to rebuild specific structs